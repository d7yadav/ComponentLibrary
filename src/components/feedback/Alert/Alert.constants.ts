import { AlertSeverity, AlertVariant, AlertSize } from './Alert.types';

/**
 * ALERT_SEVERITIES component
 * 
 * @returns JSX element
 */
export const ALERT_SEVERITIES: Record<AlertSeverity, AlertSeverity> = {
  error: 'error',
  warning: 'warning',
  info: 'info',
  success: 'success',
} as const;

/**
 * ALERT_VARIANTS component
 * 
 * @returns JSX element
 */
export const ALERT_VARIANTS: Record<AlertVariant, AlertVariant> = {
  filled: 'filled',
  outlined: 'outlined',
  standard: 'standard',
} as const;

/**
 * ALERT_SIZES component
 * 
 * @returns JSX element
 */
export const ALERT_SIZES: Record<AlertSize, AlertSize> = {
  small: 'small',
  medium: 'medium',
  large: 'large',
} as const;

/**
 * ALERT_SIZE_CONFIGS component
 * 
 * @returns JSX element
 */
export const ALERT_SIZE_CONFIGS = {
  small: {
    padding: '6px 12px',
    fontSize: '0.75rem',
    iconSize: 16,
    minHeight: 36,
    titleFontSize: '0.875rem',
    borderRadius: 6,
  },
  medium: {
    padding: '8px 16px',
    fontSize: '0.875rem',
    iconSize: 20,
    minHeight: 48,
    titleFontSize: '1rem',
    borderRadius: 8,
  },
  large: {
    padding: '12px 20px',
    fontSize: '1rem',
    iconSize: 24,
    minHeight: 64,
    titleFontSize: '1.125rem',
    borderRadius: 12,
  },
} as const;

/**
 * ALERT_COLORS component
 * 
 * @returns JSX element
 */
export const ALERT_COLORS = {
  error: {
    main: '#d32f2f',
    light: '#ffebee',
    dark: '#c62828',
    contrast: '#ffffff',
    background: '#ffebee',
    border: '#e57373',
  },
  warning: {
    main: '#ed6c02',
    light: '#fff3e0',
    dark: '#e65100',
    contrast: '#ffffff',
    background: '#fff3e0',
    border: '#ffb74d',
  },
  info: {
    main: '#0288d1',
    light: '#e3f2fd',
    dark: '#01579b',
    contrast: '#ffffff',
    background: '#e3f2fd',
    border: '#64b5f6',
  },
  success: {
    main: '#2e7d32',
    light: '#e8f5e8',
    dark: '#1b5e20',
    contrast: '#ffffff',
    background: '#e8f5e8',
    border: '#81c784',
  },
} as const;

/**
 * ALERT_ANIMATIONS component
 * 
 * @returns JSX element
 */
export const ALERT_ANIMATIONS = {
  slideIn: 'slideIn',
  fadeIn: 'fadeIn',
  scaleIn: 'scaleIn',
  bounceIn: 'bounceIn',
} as const;

/**
 * ALERT_ANIMATION_DURATIONS component
 * 
 * @returns JSX element
 */
export const ALERT_ANIMATION_DURATIONS = {
  fast: 200,
  normal: 300,
  slow: 500,
} as const;

/**
 * ALERT_ANIMATION_EASINGS component
 * 
 * @returns JSX element
 */
export const ALERT_ANIMATION_EASINGS = {
  easeInOut: 'cubic-bezier(0.4, 0, 0.2, 1)',
  easeOut: 'cubic-bezier(0.0, 0, 0.2, 1)',
  easeIn: 'cubic-bezier(0.4, 0, 1, 1)',
  spring: 'cubic-bezier(0.25, 0.46, 0.45, 0.94)',
  bounce: 'cubic-bezier(0.68, -0.55, 0.265, 1.55)',
} as const;

/**
 * ALERT_AUTO_HIDE_DURATIONS component
 * 
 * @returns JSX element
 */
export const ALERT_AUTO_HIDE_DURATIONS = {
  short: 3000,
  medium: 5000,
  long: 8000,
  persistent: 0, // Never auto-hide
} as const;

/**
 * ALERT_ELEVATION_VALUES component
 * 
 * @returns JSX element
 */
export const ALERT_ELEVATION_VALUES = {
  none: 0,
  subtle: 1,
  low: 2,
  medium: 4,
  high: 8,
  higher: 12,
  highest: 24,
} as const;

/**
 * ACCESSIBILITY_CONSTANTS component
 * 
 * @returns JSX element
 */
export const ACCESSIBILITY_CONSTANTS = {
  alertRole: 'alert',
  statusRole: 'status',
  alertDialogRole: 'alertdialog',
  regionRole: 'region',
  buttonRole: 'button',
  closeLabel: 'Close alert',
  actionLabel: 'Alert action',
  liveRegionPolite: 'polite',
  liveRegionAssertive: 'assertive',
} as const;

/**
 * ALERT_POSITIONS component
 * 
 * @returns JSX element
 */
export const ALERT_POSITIONS = {
  top: 'top',
  bottom: 'bottom',
  'top-left': 'top-left',
  'top-right': 'top-right',
  'bottom-left': 'bottom-left',
  'bottom-right': 'bottom-right',
} as const;

/**
 * ALERT_ICON_MAPPINGS component
 * 
 * @returns JSX element
 */
export const ALERT_ICON_MAPPINGS = {
  error: 'ErrorIcon',
  warning: 'WarningIcon',
  info: 'InfoIcon',
  success: 'CheckCircleIcon',
} as const;

/**
 * ALERT_MESSAGE_TEMPLATES component
 * 
 * @returns JSX element
 */
export const ALERT_MESSAGE_TEMPLATES = {
  error: {
    generic: 'An error occurred. Please try again.',
    validation: 'Please check your input and try again.',
    network: 'Network error. Please check your connection.',
    permission: 'You do not have permission to perform this action.',
    notFound: 'The requested resource was not found.',
  },
  warning: {
    generic: 'Please review the following warning.',
    unsaved: 'You have unsaved changes that may be lost.',
    deprecated: 'This feature is deprecated and will be removed.',
    quota: 'You are approaching your usage limit.',
    browser: 'Your browser may not support all features.',
  },
  info: {
    generic: 'Here is some helpful information.',
    loading: 'Loading data, please wait...',
    updated: 'Information has been updated.',
    tip: 'Pro tip: Use keyboard shortcuts for faster navigation.',
    maintenance: 'Scheduled maintenance will occur this weekend.',
  },
  success: {
    generic: 'Operation completed successfully.',
    saved: 'Your changes have been saved.',
    created: 'Item created successfully.',
    updated: 'Item updated successfully.',
    deleted: 'Item deleted successfully.',
  },
} as const;

/**
 * ALERT_PATTERNS component
 * 
 * @returns JSX element
 */
export const ALERT_PATTERNS = {
  // Common alert patterns
  form: {
    validation: {
      severity: ALERT_SEVERITIES.error,
      variant: ALERT_VARIANTS.outlined,
      closable: true,
      showIcon: true,
    },
    success: {
      severity: ALERT_SEVERITIES.success,
      variant: ALERT_VARIANTS.filled,
      autoHide: true,
      autoHideDuration: ALERT_AUTO_HIDE_DURATIONS.medium,
    },
  },
  
  page: {
    error: {
      severity: ALERT_SEVERITIES.error,
      variant: ALERT_VARIANTS.standard,
      closable: true,
      elevated: true,
      fullWidth: true,
    },
    info: {
      severity: ALERT_SEVERITIES.info,
      variant: ALERT_VARIANTS.outlined,
      closable: true,
      rounded: true,
    },
  },
  
  notification: {
    success: {
      severity: ALERT_SEVERITIES.success,
      variant: ALERT_VARIANTS.filled,
      autoHide: true,
      animated: true,
      elevated: true,
    },
    warning: {
      severity: ALERT_SEVERITIES.warning,
      variant: ALERT_VARIANTS.standard,
      closable: true,
      showIcon: true,
    },
  },
} as const;

/**
 * DEFAULT_ALERT_PROPS component
 * 
 * @returns JSX element
 */
export const DEFAULT_ALERT_PROPS = {
  severity: ALERT_SEVERITIES.info,
  variant: ALERT_VARIANTS.standard,
  size: ALERT_SIZES.medium,
  closable: false,
  showIcon: true,
  fullWidth: false,
  elevated: false,
  elevation: ALERT_ELEVATION_VALUES.none,
  rounded: false,
  centered: false,
  autoHide: false,
  autoHideDuration: ALERT_AUTO_HIDE_DURATIONS.medium,
  animated: false,
  animationDuration: ALERT_ANIMATION_DURATIONS.normal,
} as const;