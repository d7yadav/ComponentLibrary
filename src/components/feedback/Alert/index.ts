/**
 * @fileoverview Alert component barrel export
 * @author Dilip Yadav <dilip.sm.yadav@gmail.com>
 */

export { Alert } from './Alert';
export type { 
  AlertProps, 
  AlertSeverity, 
  AlertVariant,
  AlertSize,
  AlertAction,
  AlertStyleProps,
  AlertIconMapping,
  AlertColorMapping,
  AlertConfiguration
} from './Alert.types';
export {
  ALERT_SEVERITIES,
  ALERT_VARIANTS,
  ALERT_SIZES,
  ALERT_SIZE_CONFIGS,
  ALERT_COLORS,
  ALERT_ANIMATIONS,
  ALERT_ANIMATION_DURATIONS,
  ALERT_ANIMATION_EASINGS,
  ALERT_AUTO_HIDE_DURATIONS,
  ALERT_ELEVATION_VALUES,
  ACCESSIBILITY_CONSTANTS,
  ALERT_POSITIONS,
  ALERT_ICON_MAPPINGS,
  ALERT_MESSAGE_TEMPLATES,
  ALERT_PATTERNS,
  DEFAULT_ALERT_PROPS,
} from './Alert.constants';
export {
  AlertTitle,
  AlertContent,
  AlertActions,
  SlideInAlert,
  ScaleInAlert,
  BounceInAlert,
  SuccessAlert,
  ErrorAlert,
  WarningAlert,
  InfoAlert,
  CompactAlert,
  BannerAlert,
  ToastAlert,
} from './Alert.styles';