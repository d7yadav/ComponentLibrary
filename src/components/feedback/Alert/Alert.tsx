import { forwardRef, useState, useEffect, memo } from 'react';
import { IconButton } from '@mui/material';
import { 
  CheckCircle, 
  Warning, 
  Info, 
  Error as ErrorIcon,
  Close 
} from '@mui/icons-material';
import { AlertProps } from './Alert.types';
import { 
  StyledAlert, 
  AlertTitle, 
  AlertContent, 
  AlertActions 
} from './Alert.styles';
import {
  DEFAULT_ALERT_PROPS,
  ACCESSIBILITY_CONSTANTS,
  ALERT_COLORS,
} from './Alert.constants';
import { Button } from '../../core/Button';

/**
 * Enhanced Alert component with multiple severity levels and variants
 * 
 * Features:
 * - 4 severity levels: error, warning, info, success
 * - 3 variants: filled, outlined, standard
 * - Multiple sizes: small, medium, large
 * - Auto-hide functionality with customizable duration
 * - Custom actions and close button support
 * - Animation options for enhanced UX
 * - Accessibility compliance (WCAG 2.1 AA)
 * - TypeScript support with strict typing
 * - Dark theme support via enhanced theme system
 */
export const Alert = forwardRef<HTMLDivElement, AlertProps>((
  {
    children,
    severity = DEFAULT_ALERT_PROPS.severity,
    variant = DEFAULT_ALERT_PROPS.variant,
    size = DEFAULT_ALERT_PROPS.size,
    title,
    closable = DEFAULT_ALERT_PROPS.closable,
    onClose,
    icon,
    showIcon = DEFAULT_ALERT_PROPS.showIcon,
    actions,
    fullWidth = DEFAULT_ALERT_PROPS.fullWidth,
    elevated = DEFAULT_ALERT_PROPS.elevated,
    elevation = DEFAULT_ALERT_PROPS.elevation,
    rounded = DEFAULT_ALERT_PROPS.rounded,
    borderRadius,
    centered = DEFAULT_ALERT_PROPS.centered,
    bgcolor,
    color,
    autoHide = DEFAULT_ALERT_PROPS.autoHide,
    autoHideDuration = DEFAULT_ALERT_PROPS.autoHideDuration,
    animated = DEFAULT_ALERT_PROPS.animated,
    animationDuration = DEFAULT_ALERT_PROPS.animationDuration,
    className,
    sx,
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    role,
    ...other
  },
  ref
) => {
  const [visible, setVisible] = useState(true);

  // Auto-hide functionality
  useEffect(() => {
    if (autoHide && autoHideDuration > 0 && visible) {
      const timer = setTimeout(() => {
        setVisible(false);
        onClose?.();
      }, autoHideDuration);

      return () => clearTimeout(timer);
    }
  }, [autoHide, autoHideDuration, visible, onClose]);

  // Handle close
  const handleClose = () => {
    setVisible(false);
    onClose?.();
  };

  // Get default icon based on severity
  const getDefaultIcon = () => {
    const iconProps = { fontSize: 'inherit' as const };
    switch (severity) {
      case 'error':
        return <ErrorIcon {...iconProps} />;
      case 'warning':
        return <Warning {...iconProps} />;
      case 'info':
        return <Info {...iconProps} />;
      case 'success':
        return <CheckCircle {...iconProps} />;
      default:
        return <Info {...iconProps} />;
    }
  };

  // Build close action
  const closeAction = closable ? (
    <IconButton
      size="small"
      onClick={handleClose}
      aria-label={ACCESSIBILITY_CONSTANTS.closeLabel}
      sx={{ 
        color: 'inherit',
        padding: 0.5,
        marginLeft: 1,
      }}
    >
      <Close fontSize="small" />
    </IconButton>
  ) : undefined;

  // Build action buttons
  const actionButtons = actions?.length ? (
    <AlertActions>
      {actions.map((action, index) => (
        <Button
          key={index}
          variant={action.variant || 'text'}
          color={action.color || 'inherit'}
          size="small"
          onClick={action.onClick}
          aria-label={`${ACCESSIBILITY_CONSTANTS.actionLabel}: ${action.label}`}
        >
          {action.label}
        </Button>
      ))}
    </AlertActions>
  ) : null;

  // Accessibility props
  const accessibilityProps = {
    'aria-label': ariaLabel,
    'aria-describedby': ariaDescribedBy,
    role: role || (severity === 'error' ? ACCESSIBILITY_CONSTANTS.alertRole : ACCESSIBILITY_CONSTANTS.statusRole),
    'aria-live': severity === 'error' ? ACCESSIBILITY_CONSTANTS.liveRegionAssertive : ACCESSIBILITY_CONSTANTS.liveRegionPolite,
  };

  if (!visible) {
    return null;
  }

  return (
    <StyledAlert data-testid="alert"
      ref={ref}
      severity={severity}
      variant={variant}
      icon={showIcon ? (icon || getDefaultIcon()) : false}
      action={closeAction}
      className={className}
      sx={sx}
      // Custom props for styled component
      customSize={size}
      customFullWidth={fullWidth}
      customElevated={elevated}
      customElevation={elevation}
      customRounded={rounded}
      customBorderRadius={borderRadius}
      customCentered={centered}
      customBgcolor={bgcolor}
      customColor={color}
      customAnimated={animated}
      customAnimationDuration={animationDuration}
      {...accessibilityProps}
      {...other}
    >
      <div className="alert-message">
        {title && (
          <AlertTitle customSize={size} className="alert-title">
            {title}
          </AlertTitle>
        )}
        <AlertContent className="alert-content">
          {children}
        </AlertContent>
        {actionButtons}
      </div>
    </StyledAlert>
  );
});

Alert.displayName = 'Alert';