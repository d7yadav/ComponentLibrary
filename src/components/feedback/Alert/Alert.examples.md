# ðŸ“– Alert Examples

## Basic Usage

```typescript
import { Alert } from '@/components/feedback/Alert';

// Basic usage
const Example = () => {
  return (
    <Alert>
      Basic Alert example
    </Alert>
  );
};
```

## Advanced Examples

```typescript
// Advanced configuration
const AdvancedExample = () => {
  const [value, setValue] = useState('');
  
  return (
    <Alert
      variant="text"
      disabled={false}
      onChange={(e) => setValue(e.target.value)}
      sx={{
        borderRadius: 2,
        '&:hover': {
          transform: 'translateY(-2px)',
        },
      }}
    >
      Advanced Alert
    </Alert>
  );
};
```

## Integration Examples

```typescript
// Integration with form libraries
import { useController, Control } from 'react-hook-form';

interface FormIntegrationProps {
  name: string;
  control: Control;
}

const FormIntegration = ({ name, control }: FormIntegrationProps) => {
  const { field, fieldState } = useController({
    name,
    control,
    defaultValue: '',
  });

  return (
    <Alert
      {...field}
      error={!!fieldState.error}
      helperText={fieldState.error?.message}
    />
  );
};
```

## Edge Cases

```typescript
// Edge case: Empty states
const EmptyState = () => (
  <Alert disabled>
    No content available
  </Alert>
);

// Edge case: Error states
const ErrorState = () => (
  <Alert 
    error
    helperText="Something went wrong"
  />
);

// Edge case: Loading states
const LoadingState = () => (
  <Alert 
    loading
    disabled
  />
);
```

## Performance Optimization Examples

```typescript
// Memoized component for performance
const OptimizedAlert = React.memo((props: AlertProps) => {
  const memoizedValue = useMemo(() => {
    // Expensive calculation
    return heavyComputation(props.value);
  }, [props.value]);

  return (
    <Alert 
      {...props}
      value={memoizedValue}
    />
  );
});

// Lazy loading for large lists
const LazyList = () => {
  return (
    <VirtualizedList
      itemCount={1000}
      renderItem={({ index }) => (
        <Alert key={index}>
          Item {index}
        </Alert>
      )}
    />
  );
};
```

## Accessibility Examples

```typescript
// Full accessibility implementation
const AccessibleExample = () => {
  return (
    <Alert
      aria-label="Descriptive label for screen readers"
      aria-describedby="helper-text"
      role="button"
      tabIndex={0}
      onKeyDown={(e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          handleAction();
        }
      }}
    >
      Accessible Alert
    </Alert>
  );
};

// High contrast mode support
const HighContrastExample = () => {
  return (
    <Alert
      sx={{
        '@media (prefers-contrast: high)': {
          borderWidth: 2,
          borderColor: 'text.primary',
        },
      }}
    >
      High contrast support
    </Alert>
  );
};
```

---

*These examples demonstrate real-world usage patterns and best practices for the Alert component.*
