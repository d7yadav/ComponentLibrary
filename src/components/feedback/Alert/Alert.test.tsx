/**
 * @fileoverview Alert Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Alert } from './Alert';
import { DEFAULT_ALERT_PROPS } from './Alert.constants';
import type { AlertProps } from './Alert.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('Alert Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<Alert>Test content</Alert>);
      expect(screen.getByRole('status')).toBeInTheDocument();
    });

    it('renders with error severity and alert role', () => {
      renderWithTheme(<Alert severity="error">Error Alert</Alert>);
      expect(screen.getByRole('alert')).toBeInTheDocument();
    });

    it('renders children correctly', () => {
      const testContent = 'Test Alert Content';
      renderWithTheme(<Alert>{testContent}</Alert>);
      expect(screen.getByText(testContent)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-alert';
      renderWithTheme(
        <Alert data-testid={testId}>Test</Alert>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <Alert className={customClass}>Test</Alert>
      );
      expect(screen.getByRole('status')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: AlertProps['variant'][] = [
        'standard',
        'outlined',
        'filled',
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <Alert variant={variant} data-testid="variant-test">
              {variant} variant
            </Alert>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });


    describe('severity prop', () => {
      const severities: AlertProps['severity'][] = [
        'success',
        'info',
        'warning',
        'error',
      ];

      severities.forEach((severity) => {
        it(`renders ${severity} severity correctly`, () => {
          renderWithTheme(
            <Alert severity={severity} data-testid="severity-test">
              {severity} alert
            </Alert>
          );
          const element = screen.getByTestId('severity-test');
          expect(element).toBeInTheDocument();
        });
      });
    });


    it('handles closable state correctly', () => {
      const handleClose = vi.fn();
      renderWithTheme(
        <Alert closable={true} onClose={handleClose} data-testid="closable-test">
          Closable Alert
        </Alert>
      );
      
      const element = screen.getByTestId('closable-test');
      expect(element).toBeInTheDocument();
      expect(screen.getByRole('button')).toBeInTheDocument();
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('handles close button interaction', async () => {
      const user = userEvent.setup();
      const handleClose = vi.fn();
      
      renderWithTheme(
        <Alert severity="info" closable={true} onClose={handleClose} data-testid="closable-alert">
          Closable Alert
        </Alert>
      );
      
      const closeButton = screen.getByRole('button');
      await user.click(closeButton);
      expect(handleClose).toHaveBeenCalledTimes(1);
    });

    it('supports custom actions array', async () => {
      const user = userEvent.setup();
      const handleAction = vi.fn();
      
      renderWithTheme(
        <Alert 
          severity="warning"
          actions={[
            { label: 'Action 1', onClick: handleAction, variant: 'text' }
          ]}
        >
          Alert with custom actions
        </Alert>
      );
      
      const actionButton = screen.getByText('Action 1');
      await user.click(actionButton);
      expect(handleAction).toHaveBeenCalledTimes(1);
    });
  });


  // Icon testing
  describe('Icons', () => {
    const TestIcon = () => <span data-testid="test-icon">Icon</span>;

    it('renders custom icon correctly', () => {
      renderWithTheme(
        <Alert icon={<TestIcon />}>
          With Custom Icon
        </Alert>
      );
      
      expect(screen.getByTestId('test-icon')).toBeInTheDocument();
    });

    it('renders default severity icons', () => {
      renderWithTheme(
        <Alert severity="info" showIcon={true}>
          With Default Icon
        </Alert>
      );
      
      expect(screen.getByTestId('InfoIcon')).toBeInTheDocument();
    });

    it('hides icon when showIcon is false', () => {
      renderWithTheme(
        <Alert severity="info" showIcon={false}>
          Without Icon
        </Alert>
      );
      
      expect(screen.queryByTestId('InfoIcon')).not.toBeInTheDocument();
    });
  });

  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <Alert>Accessible Alert</Alert>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers', () => {
      const ariaLabel = 'Screen reader label';
      renderWithTheme(
        <Alert aria-label={ariaLabel}>
          Alert for screen readers
        </Alert>
      );
      
      expect(screen.getByLabelText(ariaLabel)).toBeInTheDocument();
    });

    it('supports keyboard navigation', () => {
      renderWithTheme(<Alert tabIndex={0}>Keyboard Alert</Alert>);
      
      const element = screen.getByRole('status');
      element.focus();
      expect(element).toHaveFocus();
    });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <Alert 
          aria-describedby="description"
          aria-controls="controlled-element"
        >
          ARIA Alert
        </Alert>
      );
      
      const element = screen.getByRole('status');
      expect(element).toHaveAttribute('aria-describedby', 'description');
      expect(element).toHaveAttribute('aria-controls', 'controlled-element');
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiAlert: {
            defaultProps: {
              variant: 'standard',
            },
            styleOverrides: {
              root: {
                backgroundColor: 'rgb(255, 0, 0)',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <Alert data-testid="themed-alert">
            Themed Alert
          </Alert>
        </ThemeProvider>
      );

      const element = screen.getByTestId('themed-alert');
      expect(element).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <Alert>Dark Mode Alert</Alert>
        </ThemeProvider>
      );

      expect(screen.getByRole('status')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyComponents = Array.from({ length: 100 }, (_, i) => (
        <Alert key={i}>Alert {i}</Alert>
      ));

      const { container } = renderWithTheme(<div>{manyComponents}</div>);
      expect(container.querySelectorAll('[role="status"]')).toHaveLength(100);
    });

  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<Alert />);
      expect(screen.getByRole('status')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(<Alert>{null}</Alert>);
      expect(screen.getByRole('status')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <Alert 
          variant={undefined as any} 
        >
          Undefined Props
        </Alert>
      );
      expect(screen.getByRole('status')).toBeInTheDocument();
    });

  });

  // Snapshot testing
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <Alert>Default Alert</Alert>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: AlertProps['variant'][] = [
        'standard',
        'outlined',
        'filled',
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <Alert variant={variant}>{variant} Alert</Alert>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});
