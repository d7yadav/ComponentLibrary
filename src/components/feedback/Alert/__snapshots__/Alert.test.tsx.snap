// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Alert Component > Snapshots > matches snapshot for all variants > variant-filled 1`] = `
<div
  aria-live="polite"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorInfo MuiAlert-filledInfo MuiAlert-filled css-s25ubi-MuiPaper-root-MuiAlert-root"
  data-testid="alert"
  role="status"
  style="--Paper-shadow: none;"
>
  <div
    class="MuiAlert-icon css-vab54s-MuiAlert-icon"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
      data-testid="InfoIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
      />
    </svg>
  </div>
  <div
    class="MuiAlert-message css-zioonp-MuiAlert-message"
  >
    <div
      class="alert-message"
    >
      <div
        class="alert-content css-qu50wt"
      >
        filled
         Alert
      </div>
    </div>
  </div>
</div>
`;

exports[`Alert Component > Snapshots > matches snapshot for all variants > variant-outlined 1`] = `
<div
  aria-live="polite"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorInfo MuiAlert-outlinedInfo MuiAlert-outlined css-1tmlh7j-MuiPaper-root-MuiAlert-root"
  data-testid="alert"
  role="status"
  style="--Paper-shadow: none;"
>
  <div
    class="MuiAlert-icon css-vab54s-MuiAlert-icon"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
      data-testid="InfoIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
      />
    </svg>
  </div>
  <div
    class="MuiAlert-message css-zioonp-MuiAlert-message"
  >
    <div
      class="alert-message"
    >
      <div
        class="alert-content css-qu50wt"
      >
        outlined
         Alert
      </div>
    </div>
  </div>
</div>
`;

exports[`Alert Component > Snapshots > matches snapshot for all variants > variant-standard 1`] = `
<div
  aria-live="polite"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorInfo MuiAlert-standardInfo MuiAlert-standard css-wx681v-MuiPaper-root-MuiAlert-root"
  data-testid="alert"
  role="status"
  style="--Paper-shadow: none;"
>
  <div
    class="MuiAlert-icon css-vab54s-MuiAlert-icon"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
      data-testid="InfoIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
      />
    </svg>
  </div>
  <div
    class="MuiAlert-message css-zioonp-MuiAlert-message"
  >
    <div
      class="alert-message"
    >
      <div
        class="alert-content css-qu50wt"
      >
        standard
         Alert
      </div>
    </div>
  </div>
</div>
`;

exports[`Alert Component > Snapshots > matches snapshot for default props 1`] = `
<div
  aria-live="polite"
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorInfo MuiAlert-standardInfo MuiAlert-standard css-wx681v-MuiPaper-root-MuiAlert-root"
  data-testid="alert"
  role="status"
  style="--Paper-shadow: none;"
>
  <div
    class="MuiAlert-icon css-vab54s-MuiAlert-icon"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
      data-testid="InfoIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
      />
    </svg>
  </div>
  <div
    class="MuiAlert-message css-zioonp-MuiAlert-message"
  >
    <div
      class="alert-message"
    >
      <div
        class="alert-content css-qu50wt"
      >
        Default Alert
      </div>
    </div>
  </div>
</div>
`;
